// benchpress/es6/dev/src/web_driver_extension.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") return Reflect.decorate(decorators, target, key, desc);
    switch (arguments.length) {
        case 2: return decorators.reduceRight(function(o, d) { return (d && d(o)) || o; }, target);
        case 3: return decorators.reduceRight(function(o, d) { return (d && d(target, key)), void 0; }, void 0);
        case 4: return decorators.reduceRight(function(o, d) { return (d && d(target, key, o)) || o; }, desc);
    }
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { bind, Injector, OpaqueToken } from 'angular2/di';
import { BaseException, ABSTRACT, isBlank } from 'angular2/src/facade/lang';
import { ListWrapper } from 'angular2/src/facade/collection';
import { Options } from './common_options';
/**
 * A WebDriverExtension implements extended commands of the webdriver protocol
 * for a given browser, independent of the WebDriverAdapter.
 * Needs one implementation for every supported Browser.
 */
export let WebDriverExtension = class {
    static bindTo(childTokens) {
        var res = [
            bind(_CHILDREN)
                .toFactory((injector) => ListWrapper.map(childTokens, (token) => injector.get(token)), [Injector]),
            bind(WebDriverExtension)
                .toFactory((children, capabilities) => {
                var delegate;
                ListWrapper.forEach(children, (extension) => {
                    if (extension.supports(capabilities)) {
                        delegate = extension;
                    }
                });
                if (isBlank(delegate)) {
                    throw new BaseException('Could not find a delegate for given capabilities!');
                }
                return delegate;
            }, [_CHILDREN, Options.CAPABILITIES])
        ];
        return res;
    }
    gc() { throw new BaseException('NYI'); }
    timeBegin(name) { throw new BaseException('NYI'); }
    timeEnd(name, restartName) { throw new BaseException('NYI'); }
    /**
     * Format:
     * - cat: category of the event
     * - name: event name: 'script', 'gc', 'render', ...
     * - ph: phase: 'B' (begin), 'E' (end), 'b' (nestable start), 'e' (nestable end), 'X' (Complete
     *event)
     * - ts: timestamp in ms, e.g. 12345
     * - pid: process id
     * - args: arguments, e.g. {heapSize: 1234}
     *
     * Based on [Chrome Trace Event
     *Format](https://docs.google.com/document/d/1CvAClvFfyA5R-PhYUmn5OOQtYMH4h6I0nSsKchNAySU/edit)
     **/
    readPerfLog() { throw new BaseException('NYI'); }
    perfLogFeatures() { throw new BaseException('NYI'); }
    supports(capabilities) { return true; }
};
WebDriverExtension = __decorate([
    ABSTRACT(), 
    __metadata('design:paramtypes', [])
], WebDriverExtension);
export class PerfLogFeatures {
    constructor({ render = false, gc = false, frameCapture = false } = {}) {
        this.render = render;
        this.gc = gc;
        this.frameCapture = frameCapture;
    }
}
var _CHILDREN = new OpaqueToken('WebDriverExtension.children');
